### PR TITLE
feat: CLI command to list the members of a Workspace [RM-388]

### PR DESCRIPTION
**Pull request by ShreyaLnuHpe**
## Ticket

[RM-388](https://hpe-aiatscale.atlassian.net/browse/RM-388)

## Description

We don't have a command that lists users/user-groups members of a Workspace and their permissions). 
For details, visit original ask- [DFR-462](https://hpe-aiatscale.atlassian.net/browse/DFR-462)

Created a new CLI command to list user/group members of a Workspace
```
% det -u admin workspace --help
usage: det workspace [-h] subcommand ...

positional arguments:
  subcommand
    help         show help for this command
    archive      archive workspace
    create       create workspace
    delete       delete workspace
    describe     describe workspace
    edit         edit workspace
    list (ls)    list all workspaces
    list-members
                 list users/user-groups members of a workspace and their roles
    list-pools   list the resource pools available to a workspace
    list-projects
                 list the projects associated with a workspace
    unarchive    unarchive workspace

```

New CLI command:

> % det -u admin workspace list-members <workspace_name>  #default: tabular
> % det -u admin workspace list-members <workspace_name> --json
> % det -u admin workspace list-members <workspace_name> --csv
> % det -u admin workspace list-members <workspace_name> --yaml

## Test Plan

No more tests needed

### Test Done
Manually Tested
E2E Test completed

Created Workspace, UserGroup, User and given Roles (though CLI/WebUI)
```
% det -u admin workspace list-members ws1
 User/Group Name   | User/Group   | Role Name
-------------------+-------+------------------
 admin             | U            | WorkspaceAdmin
 ug1               | G            | Editor
 u1                | U            | EditorRestricted
```
Similar to the Members tab in [WebUI](http://127.0.0.1:8080/det/)
Screenshot 2024-07-22 at 11 00 40 AM

Output in PGAdmin:
```
SELECT u.id AS user_id, u.username, r.role_name, u.active, u.admin
FROM users u
JOIN user_group_membership ugm ON u.id = ugm.user_id
JOIN role_assignments ra ON ugm.group_id = ra.group_id
JOIN roles r ON ra.role_id = r.id;
```
[data-1721352684733.csv](https://github.com/user-attachments/files/16300886/data-1721352684733.csv)

## Checklist

- [X] Changes have been manually QA'd
- [X] New features have been approved by the corresponding PM
- [ ] User-facing API changes have the "User-facing API Change" label
- [ ] Release notes have been added as a separate file under `docs/release-notes/`
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses have been included for new code which was copied and/or modified from any external code

[RM-388]: https://hpe-aiatscale.atlassian.net/browse/RM-388?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DFR-462]: https://hpe-aiatscale.atlassian.net/browse/DFR-462?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by hamidzr — COMMENTED** [review]
mostly lgtm thanks for working on this. i added some quick comments + a question in the slack thread.

File: harness/determined/cli/workspace.py [hunk 1]
```diff
@@ -147,6 +178,21 @@ def list_pools(args: argparse.Namespace) -> None:
...
 
 
+def list_workspace_members(args: argparse.Namespace) -> None:
+    sess = cli.setup_session(args)
+    w = api.workspace_by_name(sess, args.workspace_name)
+    resp = bindings.get_GetGroupsAndUsersAssignedToWorkspace(sess, workspaceId=w.id)
+
+    if args.json:
+        render.print_json(resp.to_json())
+    elif args.yaml:
```
Comment (varlogtim) [inline harness/determined/cli/workspace.py:188]: Nice. I like that there is a yaml option even though not all other CLI items have this.

**Review by hamidzr — APPROVED** [review]
I'd check the way we mark users and groups with product in case they have a preference otherwise the functionality lgtm. maybe the header could be full `User/Group` to make it more clear.

Also, would we want to trim the `user` and `permission` objects from the json/yaml outputs? Is that something we do with any of our other CLIs? Let's skip this unless we get confirmation we want this. 

```
.usersAssignedDirectly[]
.groups[].users[]
.assignments[].role.permissions[]


  role:
    name: WorkspaceAdmin
    permissions:
        # just keep the ids?
        - id: PERMISSION_TYPE_VIEW_TEMPLATES
          name: view templates
          scopeTypeMask:
            cluster: true
            workspace: true
    roleId: 2
    scopeTypeMask:
      cluster: true
      workspace: true



- active: true
  admin: false
  agentUserGroup: null
  displayName: ''
  id: 2
  lastAuthAt: null
  modifiedAt: '2024-07-22T16:02:07.363788Z'
  remote: false
  # just keep the username?
  username: determined
```

File: e2e_tests/tests/cluster/test_workspace_org.py [hunk 2]
```diff
@@ -588,3 +589,66 @@ def test_workspaceid_set() -> None:
...
+    try:
+        admin_sess = api_utils.admin_session()
+
+        # Add a test workspace.
+        workspace_name = api_utils.get_random_string()
+        create_workspace_cmd = ["det", "workspace", "create", workspace_name]
+        detproc.check_call(admin_sess, create_workspace_cmd)
+
+        # Create test users (3) and groups (2) and assign to workspace.
+        roles = ["Editor", "Viewer", "WorkspaceAdmin", "EditorRestricted", "ModelRegistryViewer"]
```
Comment (ShreyaLnuHpe) [inline e2e_tests/tests/cluster/test_workspace_org.py:611]: Do we have some `enum` defined or a better way to access them? I couldn't find.

File: harness/determined/cli/workspace.py [hunk 3]
```diff
@@ -46,6 +46,37 @@ def get_workspace_names(session: api.Session) -> Dict[int, str]:
...
+) -> None:
+    group_map = {}
+    user_map = {}
+    values = []
+
+    if resp.groups.count != 0:
+        for g in resp.groups:
+            group_map[g.groupId] = g.name
+
+    if resp.usersAssignedDirectly.count != 0:
```
[the remark this answers is not in the document]
Comment (ShreyaLnuHpe) [reply harness/determined/cli/workspace.py]: Corrected

File: harness/determined/cli/workspace.py [hunk 4]
```diff
@@ -420,10 +424,16 @@ def yaml_file_arg(val: str) -> Any:
...
                 [
                     cli.Arg("workspace_name", type=str, help="name of the workspace"),
-                    cli.Arg("--json", action="store_true", help="print as JSON"),
+                    cli.Group(
+                        cli.output_format_args["json"],
+                        cli.output_format_args["yaml"],
+                        cli.output_format_args["table"],
+                        cli.output_format_args["csv"],
+                    ),
+                    *cli.make_pagination_args(limit=1000),
```
[the remark this answers is not in the document]
Comment (ShreyaLnuHpe) [reply harness/determined/cli/workspace.py]: Removed it

File: e2e_tests/tests/cluster/test_workspace_org.py (hunk 2 above)
Comment (hamidzr) [reply e2e_tests/tests/cluster/test_workspace_org.py:611]: not that I know of. but another suggestion around here would be to use this utility instead which directly uses our apis to set up (i think) what you're tying to set up https://github.com/determined-ai/determined/blob/769d600150439baca9f920ba4057f056d7d7a3d4/e2e_tests/tests/cluster/test_rbac.py#L25

**Review by hamidzr — APPROVED** [review]

File: e2e_tests/tests/cluster/test_workspace_org.py (hunk 2 above)
Comment (ShreyaLnuHpe) [reply e2e_tests/tests/cluster/test_workspace_org.py:612]: I would prefer keeping the code same to maintain the consistency while creating the group, their roles and assigning to the workspace.
Thanks for suggesting though.

File: harness/determined/cli/workspace.py [hunk 5]
```diff
@@ -501,6 +542,20 @@ def yaml_file_arg(val: str) -> Any:
...
+            cli.Cmd(
+                "list-members",
+                list_workspace_members,
+                "list users/user-groups members of a workspace and their roles",
+                [
+                    cli.Arg("workspace_name", type=str, help="name of the workspace"),
+                    cli.Group(
+                        cli.output_format_args["json"],
+                        cli.output_format_args["yaml"],
+                        cli.output_format_args["table"],
```
Comment (azhou-determined) [inline harness/determined/cli/workspace.py:554]: let's not include "--table", it's the default and not a "real" format

**Review by azhou-determined — APPROVED** [review]
one comment that's not necessarily blocking, but otherwise LGTM!

**Pull request merged**